### PR TITLE
feature/add prometheus metrics

### PR DESCRIPTION
**Pull request by jmcconnell26**
Hi! 

We've been looking to use smokescreen in production, and want to expose prometheus metrics to integrate with our monitoring stack.
I've raised this PR to make the changes required to enable exposing the existing metrics collected via prometheus.
Please let me know if there are any changes you'd like me to make!

Thanks,
Josh

### CODE REVIEW

**Comment by cds2-stripe** [comment]
Hey @jmcconnell26 ! Apologies for the delay here.

Is this still something you're interested in pursuing? If so I can bring this up at our team meeting. We're actually looking at moving to Prometheus internally so this is very timely.

**Comment by jmcconnell26** [comment]
Hi @cds2-stripe, yes this is something we would definitely be keen to push forward with!

There are one or two changes we made to our fork which need to be synced with this branch, but I can hopefully get those made at some point this week or next.

Thanks!
Josh

**Comment by jmcconnell26** [comment]
Hi @cds2-stripe, I've merged the changes we had been using internally. If there's anything else you'd like me to change, please just let me know!

Thanks,
Josh

**Comment by alexmv** [comment]
We'd love this feature!  Anything we can do to help it get reviewed and merged?

**Comment by cds2-stripe** [comment]
I'm reviewing and testing this now, incredibly sorry for the delays. @jmcconnell26 would you be willing to fix up the merge conflicts?

**Review by cds2-stripe — CHANGES_REQUESTED** [review]
Overall this looks great. I don't have any material feedback on the code, thanks for your tests and having the Prometheus client match the style / format of the existing statsd client.

I will need to do some more testing internally. We have some very high RPS endpoints using `IncrWithTags` and I'm curious to see any perf implications with the new usage of `constructTagArray`.

If you could fixup the merge conflict I can work on testing this in our QA environment.

**Comment by jmcconnell26** [comment]
Hi @cds2-stripe, of course! I've just updated the code to resolve the merge conflicts now. Just let me know if there are any more changes you want me to make!

File: pkg/smokescreen/metrics/metrics.go [hunk 1]
```diff
@@ -43,143 +39,18 @@ var metrics = []string{
...
 	Incr(string, float64) error
-	IncrWithTags(string, []string, float64) error
+	IncrWithTags(string, map[string]string, float64) error
 	Gauge(string, float64, float64) error
 	Histogram(string, float64, float64) error
-	HistogramWithTags(string, float64, []string, float64) error
+	HistogramWithTags(string, float64, map[string]string, float64) error
 	Timing(string, time.Duration, float64) error
-	TimingWithTags(string, time.Duration, float64, []string) error
-	StatsdClient() statsd.ClientInterface
```
Comment (cds2-stripe) [inline pkg/smokescreen/metrics/metrics.go:65]: We wrap smokescreen internal with some of our other observability tooling, and we need to retain access to the underlying statsd client to pass to other internal libraries.

What would you think about keeping this in the interface (in a perfect world it wouldn't be there, but this has been there for a while) and returning nil from the prometheus client?

**Review by cds2-stripe — CHANGES_REQUESTED** [review]

**Review by cds2-stripe — APPROVED** [review]
@jmcconnell26 I've tested this internally without any issues. We really appreciate the PR!

**Comment by cds2-stripe** [comment]
Looks like `go vet -v ./...` is failing: https://github.com/stripe/smokescreen/actions/runs/5011619656/jobs/8997700754?pr=179

**Comment by jmcconnell26** [comment]
@cds2-stripe - apologies, I've updated, and hopefully fixed the issue with go vet

**Review by cds2-stripe — APPROVED** [review]

**Comment by cds2-stripe** [comment]
Thanks again @jmcconnell26. Feel free to followup with another PR if you're interested in adding yourself to the contributors list - I'd be happy to approve!

**Comment by jmcconnell26** [comment]
Fantastic, many thanks @cds2-stripe for the review!

**Pull request merged**